# Test for Fixed Effects

## Load Libraries and Dataset

### Libraries

In [30]:
import numpy as np
import pandas as pd
import scipy.stats as stats
import statsmodels.api as sm
from linearmodels.panel import PanelOLS
from linearmodels.panel import PooledOLS

### Dataset and PanelOLS Object

In [31]:
sgx = pd.read_csv("data/clean_sgx.csv", index_col= None)
sgx = sgx[['Company Code', 'Year', 'LEVERAGE',
           'SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV']]
sgx['year'] = pd.Categorical(sgx['Year'])
sgx['compcode'] = pd.Categorical(sgx['Company Code'])
sgx = sgx.set_index(['Company Code', 'Year'])

endo = sgx.LEVERAGE
exog_vars = ['SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV']
exog = sm.add_constant(sgx[exog_vars])

pooled_mod = PooledOLS(endo, exog)
pooled_res = pooled_mod.fit()

## F-test for Fixed Effects

In [28]:
def fixedeffect(endo = endo, exog = exog,
                entity_eff = True, time_eff = True):
    mod = PanelOLS(endo, exog,
                   entity_effects=entity_eff,
                   time_effects=time_eff)
    res = mod.fit()

    return res

### Joint test for two-way FE

In [32]:
fixedeffect().summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.2294
Estimator:                   PanelOLS   R-squared (Between):              0.1387
No. Observations:                1728   R-squared (Within):               0.2397
Date:                Sun, Mar 31 2024   R-squared (Overall):              0.1569
Time:                        17:15:03   Log-likelihood                    1628.0
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      74.386
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                  F(6,1499)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             74.386
                                        P-value                           0.0000
Time periods:                       8   Distribution:                  F(6,1499)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.5191     0.2083    -7.2943     0.0000     -1.9276     -1.1106
SIZE              0.0816     0.0106     7.7306     0.0000      0.0609      0.1024
PROFITABILITY    -0.2288     0.0365    -6.2647     0.0000     -0.3004     -0.1572
TANG              0.4097     0.0369     11.103     0.0000      0.3373      0.4821
LIQUID           -0.0228     0.0022    -10.314     0.0000     -0.0272     -0.0185
MCAP          -3.265e-12  4.142e-12    -0.7884     0.4306  -1.139e-11   4.858e-12
SOLV             -0.0002  5.632e-05    -3.2463     0.0012     -0.0003  -7.236e-05
=================================================================================

F-test for Poolability: 25.310
P-value: 0.0000
Distribution: F(222,1499)

Included effects: Entity, Time
"""

Joint F-statistic = F-statistic (robust) = **74.386** with p-value smaller than 0.0001 (99.99% confidence level).

Highly significant that the two-way fixed effect exists compared to the PooledOLS null hypothesis.

### Conditional test for entity FE

In [22]:
fixedeffect(entity_eff=True, time_eff=False).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.2401
Estimator:                   PanelOLS   R-squared (Between):              0.1308
No. Observations:                1728   R-squared (Within):               0.2401
Date:                Sun, Mar 31 2024   R-squared (Overall):              0.1504
Time:                        16:33:13   Log-likelihood                    1615.1
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      79.286
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                  F(6,1506)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             79.286
                                        P-value                           0.0000
Time periods:                       8   Distribution:                  F(6,1506)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.5595     0.2029    -7.6857     0.0000     -1.9575     -1.1615
SIZE              0.0834     0.0103     8.0947     0.0000      0.0632      0.1036
PROFITABILITY    -0.2270     0.0364    -6.2424     0.0000     -0.2984     -0.1557
TANG              0.4346     0.0365     11.909     0.0000      0.3630      0.5061
LIQUID           -0.0228     0.0022    -10.308     0.0000     -0.0271     -0.0184
MCAP          -4.195e-12  4.137e-12    -1.0140     0.3107  -1.231e-11    3.92e-12
SOLV             -0.0002  5.645e-05    -3.4696     0.0005     -0.0003  -8.513e-05
=================================================================================

F-test for Poolability: 25.763
P-value: 0.0000
Distribution: F(215,1506)

Included effects: Entity
"""

F-statistic = **79.286** for conditional entity effect.

### Conditional test for time FE

In [23]:
fixedeffect(entity_eff=False, time_eff=True).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.3616
Estimator:                   PanelOLS   R-squared (Between):              0.4295
No. Observations:                1728   R-squared (Within):               0.0537
Date:                Sun, Mar 31 2024   R-squared (Overall):              0.3621
Time:                        16:37:08   Log-likelihood                    285.77
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      161.80
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                  F(6,1714)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             161.80
                                        P-value                           0.0000
Time periods:                       8   Distribution:                  F(6,1714)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -0.5893     0.0629    -9.3647     0.0000     -0.7127     -0.4659
SIZE              0.0419     0.0032     13.245     0.0000      0.0357      0.0481
PROFITABILITY    -0.4509     0.0585    -7.7111     0.0000     -0.5656     -0.3362
TANG              0.1898     0.0249     7.6117     0.0000      0.1409      0.2387
LIQUID           -0.0547     0.0027    -19.903     0.0000     -0.0601     -0.0493
MCAP          -8.712e-12  1.851e-12    -4.7055     0.0000  -1.234e-11  -5.081e-12
SOLV             -0.0007  7.056e-05    -9.5412     0.0000     -0.0008     -0.0005
=================================================================================

F-test for Poolability: 1.0642
P-value: 0.3842
Distribution: F(7,1714)

Included effects: Time
"""

F-statistic = **161.80** for conditional time effect.

### Marginal test for entity FE

In [26]:
exog_vars_marginal_entity = ['SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV', 'year']
exog_marginal_entity = sm.add_constant(sgx[exog_vars_marginal_entity])
fixedeffect(exog = exog_marginal_entity, time_eff = False).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.2513
Estimator:                   PanelOLS   R-squared (Between):              0.1387
No. Observations:                1728   R-squared (Within):               0.2513
Date:                Sun, Mar 31 2024   R-squared (Overall):              0.1589
Time:                        16:41:44   Log-likelihood                    1628.0
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      38.706
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                 F(13,1499)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             38.706
                                        P-value                           0.0000
Time periods:                       8   Distribution:                 F(13,1499)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.5240     0.2074    -7.3469     0.0000     -1.9309     -1.1171
SIZE              0.0816     0.0106     7.7306     0.0000      0.0609      0.1024
PROFITABILITY    -0.2288     0.0365    -6.2647     0.0000     -0.3004     -0.1572
TANG              0.4097     0.0369     11.103     0.0000      0.3373      0.4821
LIQUID           -0.0228     0.0022    -10.314     0.0000     -0.0272     -0.0185
MCAP          -3.265e-12  4.142e-12    -0.7884     0.4306  -1.139e-11   4.858e-12
SOLV             -0.0002  5.632e-05    -3.2463     0.0012     -0.0003  -7.236e-05
year.2016.0       0.0015     0.0098     0.1587     0.8739     -0.0176      0.0207
year.2017.0      -0.0068     0.0098    -0.6993     0.4845     -0.0260      0.0124
year.2018.0       0.0031     0.0098     0.3197     0.7492     -0.0160      0.0223
year.2019.0       0.0305     0.0099     3.0813     0.0021      0.0111      0.0499
year.2020.0       0.0171     0.0099     1.7181     0.0860     -0.0024      0.0366
year.2021.0      -0.0029     0.0100    -0.2934     0.7693     -0.0225      0.0166
year.2022.0      -0.0030     0.0099    -0.3042     0.7610     -0.0225      0.0164
=================================================================================

F-test for Poolability: 25.991
P-value: 0.0000
Distribution: F(215,1499)

Included effects: Entity
"""

F-statistic = **38.706** for marginal entity effect.

### Marginal test for time FE

In [27]:
exog_vars_marginal_time = ['SIZE', 'PROFITABILITY', 'TANG', 'LIQUID', 'MCAP', 'SOLV', 'compcode']
exog_marginal_time = sm.add_constant(sgx[exog_vars_marginal_time])
fixedeffect(exog = exog_marginal_time, entity_eff = False).summary

<class 'linearmodels.compat.statsmodels.Summary'>
"""
                          PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:               LEVERAGE   R-squared:                        0.8650
Estimator:                   PanelOLS   R-squared (Between):              1.0000
No. Observations:                1728   R-squared (Within):               0.2397
Date:                Sun, Mar 31 2024   R-squared (Overall):              0.8636
Time:                        16:43:40   Log-likelihood                    1628.0
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      43.449
Entities:                         216   P-value                           0.0000
Avg Obs:                       8.0000   Distribution:                F(221,1499)
Min Obs:                       8.0000                                           
Max Obs:                       8.0000   F-statistic (robust):             43.449
                                        P-value                           0.0000
Time periods:                       8   Distribution:                F(221,1499)
Avg Obs:                       216.00                                           
Min Obs:                       216.00                                           
Max Obs:                       216.00                                           
                                                                                
                               Parameter Estimates                               
=================================================================================
               Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
---------------------------------------------------------------------------------
const            -1.7630     0.1957    -9.0065     0.0000     -2.1470     -1.3790
SIZE              0.0816     0.0106     7.7306     0.0000      0.0609      0.1024
PROFITABILITY    -0.2288     0.0365    -6.2647     0.0000     -0.3004     -0.1572
TANG              0.4097     0.0369     11.103     0.0000      0.3373      0.4821
LIQUID           -0.0228     0.0022    -10.314     0.0000     -0.0272     -0.0185
MCAP          -3.265e-12  4.142e-12    -0.7884     0.4306  -1.139e-11   4.858e-12
SOLV             -0.0002  5.632e-05    -3.2463     0.0012     -0.0003  -7.236e-05
compcode.10       0.4267     0.0511     8.3549     0.0000      0.3265      0.5269
compcode.12       0.6185     0.0533     11.605     0.0000      0.5140      0.7231
compcode.14       0.3194     0.0541     5.9004     0.0000      0.2132      0.4256
compcode.19      -0.0147     0.0568    -0.2598     0.7951     -0.1261      0.0966
compcode.27       0.4703     0.0556     8.4517     0.0000      0.3612      0.5795
compcode.28       0.6310     0.0523     12.074     0.0000      0.5285      0.7335
compcode.29       0.3182     0.0520     6.1148     0.0000      0.2162      0.4203
compcode.33       0.4563     0.0509     8.9628     0.0000      0.3564      0.5562
compcode.34       0.4371     0.0558     7.8378     0.0000      0.3277      0.5465
compcode.47       0.4762     0.0613     7.7678     0.0000      0.3560      0.5965
compcode.49       0.3215     0.0551     5.8377     0.0000      0.2134      0.4295
compcode.50       0.3799     0.0585     6.4888     0.0000      0.2651      0.4947
compcode.52       0.6860     0.0586     11.696     0.0000      0.5709      0.8010
compcode.54       0.0952     0.0509     1.8703     0.0616     -0.0046      0.1951
compcode.57       0.5065     0.0522     9.7011     0.0000      0.4041      0.6089
compcode.64       0.2215     0.0509     4.3488     0.0000      0.1216      0.3214
compcode.65       0.2178     0.0595     3.6590     0.0003      0.1011      0.3346
compcode.67       0.3595     0.0521     6.9007     0.0000      0.2573      0.4617
compcode.68       0.3468     

F-statistic = **43.449** for marginal entity effect.

## LR test for Restricted and Unrestricted models

In [33]:
# Joint LR-test
unres_loglikehood = fixedeffect().loglik
res_loglikehood = pooled_res.loglik

lr_ratio = -2*(res_loglikehood - unres_loglikehood)
lr_df = pooled_res.df_model
lr_pval = 1 - stats.chi2.cdf(lr_ratio, lr_df)

print('LR test for 2FE')
print(f'Log-likelihood Ratio: {round(lr_ratio, 4)}')
print(f'p-value: {round(lr_pval, 4)}')
print(f'Distribution: chisq({lr_df})')

LR test for 2FE
Log-likelihood Ratio: 2691.9046
p-value: 0.0
Distribution: chisq(7)
